core/include/scipp/core/element/rebin.h

```diff
@@ -37,13 +60,19 @@ static constexpr auto rebin = overloaded{
           const auto delta = std::min<double>(xn_high, xo_high) -
                              std::max<double>(xn_low, xo_low);
           const auto owidth = xo_high - xo_low;
-          const auto scale = delta / owidth;
-          if constexpr (is_ValueAndVariance_v<
-                            std::decay_t<decltype(data_old)>>) {
+
+          constexpr bool isValAndVariance =
+              is_ValueAndVariance_v<std::decay_t<decltype(data_old)>>;
+
+          if constexpr (isValAndVariance) {
+            // Avoid implicit promotion and demotion when multiplying by scale
+            const auto scale =
+                std::remove_reference_t<decltype(data_new.value[0])>(delta /
+                                                                     owidth);
             data_new.value[inew] += data_old.value[iold] * scale;
             data_new.variance[inew] += data_old.variance[iold] * scale;
           } else {
-            data_new[inew] += data_old[iold] * scale;
+            rebinPrimativeType(data_new, data_old, inew, iold, delta, owidth);
```

Comment:
```suggestion
            if constexpr(std::is_same_v<decltype(data_new[0]), bool)
              data_new[inew] |= data_old[iold];
            else
              data_new[inew] += data_old[iold] * scale;
```
?

Reply:
👍 - this was my original implementation, but then we had mixed types (bool, double) to still handle according to the compiler

I'll add a static_assert to guarantee they use the same type and use this form since I prefer this too

Reply:
This must be a bug in how it is used. There should *not* be mixed types showing up here. Don't add hacks here, we should try to find out where the actual problem lies.